File: masterclass/Demomvp/app/src/main/java/com/example/demo_mvp/home/HomePresenter.kt
package com.example.demo_mvp.home

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancelChildren
import kotlinx.coroutines.launch

/**
 * Mediator between View and Model. No Android Context, no widget references —
 * this is what makes the Presenter unit-testable with a fake View (see
 * references/ui-pattern-mvp.md in the android-development skill).
 */
class HomePresenter(
    private val model: HomeModel,
) : HomeContract.Presenter {

    private var view: HomeContract.View? = null
    private val presenterScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    override fun attachView(view: HomeContract.View) {
        this.view = view
    }

    override fun detachView() {
        presenterScope.coroutineContext.cancelChildren()
        view = null
    }

    override fun loadData() {
        view?.showLoading()
        presenterScope.launch {
            try {
                val data = model.fetchHomeData()
                view?.hideLoading()
                view?.showData(data)
            } catch (e: Exception) {
                view?.hideLoading()
                view?.showError(e.message ?: "Unknown error")
            }
        }
    }

    override fun onTransactionClicked(transactionId: String) {
        presenterScope.launch { model.markTransactionSeen(transactionId) }
    }
}

File: masterclass/Demo-mvp-iOS/Demo-mvp-iOS/Home/HomePresenter.swift
import Foundation

@MainActor
final class HomePresenter: HomePresenterProtocol {

    private let model: HomeModel
    private weak var view: HomeViewProtocol?
    private var tasks: [Task<Void, Never>] = []

    init(model: HomeModel) {
        self.model = model
    }

    func attachView(_ view: HomeViewProtocol) {
        self.view = view
    }

    func detachView() {
        tasks.forEach { $0.cancel() }
        tasks.removeAll()
        view = nil
    }

    func loadData() {
        view?.showLoading()
        tasks.append(Task { [weak self] in
            guard let self else { return }
            do {
                let data = try await model.fetchHomeData()
                view?.hideLoading()
                view?.showData(data)
            } catch {
                view?.hideLoading()
                view?.showError("No se pudo cargar la información")
            }
        })
    }

    func onTransactionClicked(_ transactionId: String) {
        tasks.append(Task { [model] in
            await model.markTransactionSeen(transactionId: transactionId)
        })
    }
}
